changefeedccl: added protected time stamp age to monitoring metrics.

surahman opened the pull request:
Release note (ccl change):
The protected time stamp's age is now available as a monitoring metric (gauge).

Closes #82825 
Jira issue: [CRDB-16687](https://cockroachlabs.atlassian.net/browse/CRDB-16687)

CC @miretskiy

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/83013)
<!-- Reviewable:end -->

miretskiy commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1272,6 +1272,7 @@ func (cf *changeFrontier) manageProtectedTimestamps(
 		return nil
 	}
 	cf.lastProtectedTimestampUpdate = timeutil.Now()
+	cf.metrics.ProtectedTimeStamp = cf.lastProtectedTimestampUpdate
```
This is not safe: metrics are global and there could be multiple changefeeds running
on the server.

surahman replied:
If there is a shared metric between multiple change feeds and each feed can be attached to a unique ID:

```go
ProtectedTimeStamp struct {
    syncutil.Mutex
    TimeStamps map[int]time.Time
}
```

Use the oldest change feed time stamp as the data for the gauge.

miretskiy reviewed (changes requested):
I need to think a bit more about the best way to keep track of this data.

abarganier reviewed (commented):
Just one nit from a broader o11y perspective. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83013#-:-N567IFa3EcpOFUy1Hay:bnsni3y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @surahman)

___
*[`pkg/ccl/changefeedccl/metrics.go` line 328 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/83013#-N566EY_EBgMRnYGjm7Q:-N566EY_EBgMRnYGjm7R:bnopk4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ac080ad208dc742273d1a50739c36e1b3278f2d6/pkg/ccl/changefeedccl/metrics.go#L328)):*
> ```Go
> 	metaProtectedTimeStampBehindNanos = metric.Metadata{
> 		Name:        "changefeed.protected_time_stamp_behind_nanos",
> 		Help:        "The amount of time since the protected stamp has been updated.",
> ```

nit: can you make this a bit more descriptive? What is the significance of this timestamp to somebody who might need this to debug? What might a high value indicate? What about one that's constantly low? Anything that can de-mystify this metric would be very helpful to future users/maintainers.

Additionally, if this is a metric you might want to setup an alert on, consider adding a related [alerting rule](https://github.com/cockroachdb/cockroach/blob/master/pkg/util/metric/rule.go#L40-L58)



_Code quote:_
```Go
Help:        "The amount of time since the protected stamp has been updated.",
```


<!-- Sent from Reviewable.io -->

surahman commented:
> nit: can you make this a bit more descriptive? What is the significance of this timestamp to somebody who might need this to debug? What might a high value indicate? What about one that's constantly low? Anything that can de-mystify this metric would be very helpful to future users/maintainers.

From my understanding after reviewing the code, the PTS indicates how far behind the change feed is. The larger the value the more likely that new data has not propagated. It is a staleness/freshness issue.

> Additionally, if this is a metric you might want to setup an alert on, consider adding a related [alerting rule](https://github.com/cockroachdb/cockroach/blob/master/pkg/util/metric/rule.go#L40-L58)

I believe there would need to be an addition in the [KVServer Metrics](https://github.com/cockroachdb/cockroach/blob/75cf2bb581fdc57db0928d0e907d94003e760c04/pkg/kv/kvserver/metric_rules.go) for this.

I am waiting on input as to how best to track the multiple PTS that sit behind this metric. It might be reasonable to use a map to store the PTS per change feed and report the oldest using a gauge. This would indicate how old the changes might be. A histogram could be used if there is a need to track all the change feeds.

abarganier reviewed (commented):
> From my understanding after reviewing the code

Thanks for the description - my main feedback here is that ideally, the description is clear enough where people don't have to *read* the code to understand it's purpose/measurement. I don't think this is the case in its current state. The description you gave was very helpful. Specifically, the parts about what larger values might mean & how it may indicate a staleness/freshness issue, would be nice to include in the description in some way.

No worries if we are waiting on how to best track, I don't have any opinions on that (as long as the description is clear! :smile:) 

> I believe there would need to be an addition in the KVServer Metrics for this.

SGTM! Thanks for putting it on your radar. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/83013#-:-N56m_lr0o0jwaGilKhf:b-mxhn9r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @miretskiy, and @surahman)



<!-- Sent from Reviewable.io -->

surahman commented:
@abarganier all of your suggestions are great, and I will affect them. The alerting is important because stale data are symptomatic of a broader set of issues like node and network issues/failures. In exceptional cases, it could be caused by a network partition.

surahman commented:
From the CI logs, the race condition inspector seems to have been triggered - as expected. I had a quick look at the issue:

The `id` is protected by the mutex

https://github.com/cockroachdb/cockroach/blob/8601bedf62a7e0534a99c33087cdafcc6d1015fd/pkg/ccl/changefeedccl/metrics.go#L506-L507

appears to be a simple monotonic unique ID token generator (integer) that is used to set the `cf.metricsID`

https://github.com/cockroachdb/cockroach/blob/8601bedf62a7e0534a99c33087cdafcc6d1015fd/pkg/ccl/changefeedccl/changefeed_processors.go#L720-L722

as observed here

https://github.com/cockroachdb/cockroach/blob/8601bedf62a7e0534a99c33087cdafcc6d1015fd/pkg/ccl/changefeedccl/changefeed_processors.go#L973-L977

and it is never decremented on tear-down, as observed here

https://github.com/cockroachdb/cockroach/blob/8601bedf62a7e0534a99c33087cdafcc6d1015fd/pkg/ccl/changefeedccl/changefeed_processors.go#L1006-L1018

This single mutex is used for all concurrent operations in the `metrics.Metrics` struct and is only accessed within `changefeed_processor`. Operations protected by the mutex are not interleaved.

If `cf.metricsID` is a unique and can be tied to a PTS:

1. New struct with a mutex for the token generator `metrics.Metrics.mu.id`
2. New struct with mutex for `metrics.Metrics.mu.resolved`
3. New struct with mutex for the change feed PTS

Care will need to be taken to ensure there are no nested locking calls (potential deadlock). I do not believe this will be an issue as none of the current logic would require lock nesting. This may also improve concurrency.

miretskiy commented:
@surahman I'm sorry I've been a bit swamped; I'll get back to this pr, hopefully soon.

surahman commented:
@miretskiy thank you, your input is greatly appreciated - there is no rush. In the meantime, I would like to run a quick build test with my idea.

surahman commented:
Hello @abarganier, how to correctly track the metric is still in flux but I do however need to configure the PromQL expression for the alert. To that effect, I would need a threshold value from `O11y` in `ns` to exceed before triggering. 

The alerting rules for Prometheus as well as the Graphana dashboard will need to be configured respectively:
1. [alert.rules.yml](https://github.com/cockroachdb/cockroach/blob/9e47dc901dbc901609945f5241a58c318c64f5f9/monitoring/rules/alerts.rules.yml)
2. [changefeeds.json](https://github.com/cockroachdb/cockroach/blob/9e47dc901dbc901609945f5241a58c318c64f5f9/monitoring/grafana-dashboards/changefeeds.json)

I am seeing a 404 error when I access the CRDB console and empty dashboards in Graphana when I launch the [docker containers](https://github.com/cockroachdb/cockroach/blob/9e47dc901dbc901609945f5241a58c318c64f5f9/monitoring/README.md) for metrics configuration.

surahman committed:
changefeedccl: added protected time stamp age to monitoring metrics.

Release note (ccl change):
The protected time stamp's age is now available as a monitoring metric (gauge). There is now also
an alert for when the time stamp age exceeds a threshold.
